Polars: add strptime

Shoeboxam opened the pull request:
Towards #1755

Shoeboxam commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/2096?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2096** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2096?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#2095** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2095?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#2091** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2091?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#2089** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2089?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

mccalluc commented on rust/src/domains/polars/datetime/mod.rs:
```diff
@@ -0,0 +1,18 @@
+#[derive(Debug, Clone, PartialEq)]
+pub struct DatetimeDomain {
+    pub time_unit: TimeUnit,
+    pub time_zone: Option<String>,
```
This seems odd, but I'm not sure I understand the goal. Expected a composition of Date and Time. Are we expecting to get anything out of the timezone string? Should there be hours ahead or behind UTC?

Shoeboxam replied:
* Time (i64): number of nanoseconds since midnight
* Datetime (i64): number of milli/nano/micro seconds since Unix epoch. Time zone and time unit are encoded as part of the type/shared amongst all rows in a column
* Date (i32): number of days since Unix epoch.

Time and Date don't carry any extra type information/domain descriptors, so they can be represented via AtomDomain.
Datetime is separate as it isn't a composition of date and time, it is its own data format. It is also different in that it has type information (zone and unit) that needs to be known to interpret the meaning of the data. For the purposes of OpenDP, I consider the set of time points spanned by the domain to be distinct for each time zone. Updated the documentation.

mccalluc commented on rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs:
```diff
@@ -0,0 +1,107 @@
+    }
+    // Strict casting makes the transformation unstable: errors tell you things about private data.
+    // Could throw an error if strict, but it is the default, so for ease-of-use it is forced to be non-strict.
+    // It is also ok for overflow to wraparound.
```
What is the range of dates which we can represent? If dates earlier than 1970 wrap around, that is going to be undesirable in many cases.

Shoeboxam replied:
Dates and datetimes are signed offsets from the epoch, so dates earlier than 1970 don't wrap until much earlier time points.

mccalluc commented on rust/src/transformations/make_stable_expr/expr_str/expr_strptime/mod.rs:
```diff
@@ -0,0 +1,107 @@
+        .ok()
+        .flatten()
+        .unwrap_or("raise".to_string());
+    let ambiguous = if ["earliest".to_string(), "latest".to_string()].contains(&ambiguous_str) {
```
"earliest" and "latest" don't occur elsewhere in this PR. Are they coming from a library?

mccalluc replied:
Never mind: found https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.str.strptime.html

mccalluc commented on rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs:
```diff
@@ -0,0 +1,87 @@
+                },
+                lit("null"),
+            )
+            .alias("date"),
```
Do you need the alias if the name doesn't change?

mccalluc commented on rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs:
```diff
@@ -0,0 +1,87 @@
+
+    let actual = t_casted.invoke(&lf)?.collect()?;
+    let expected = r#"shape: (4, 3)
+┌──────────┬─────────────────────┬────────────┐
+│ time     ┆ datetime            ┆ date       │
+│ ---      ┆ ---                 ┆ ---        │
+│ time     ┆ datetime[ns]        ┆ date       │
+╞══════════╪═════════════════════╪════════════╡
+│ 14:00:00 ┆ 2023-08-12 14:00:00 ┆ 2023-08-12 │
+│ 15:00:00 ┆ 2024-09-13 15:00:00 ┆ 2024-09-13 │
+│ 16:00:00 ┆ 2025-10-14 16:00:00 ┆ 2025-10-14 │
+│ 16:00:00 ┆ null                ┆ null       │
+└──────────┴─────────────────────┴────────────┘"#;
```
This kind of test is going to be fragile if the data changes... but it might be easier than constructing a dataframe to check against.

mccalluc commented on rust/src/transformations/make_stable_expr/expr_str/expr_strptime/test.rs:
```diff
@@ -0,0 +1,87 @@
+    ])?;
+    let lf = df!(
+        "time" => &[
+            "14:00:00".to_string(),
+            "15:00:00".to_string(),
+            "16:00:00".to_string(),
+            "16:00:00".to_string()
+        ],
+        "datetime" => &[
+            "2023-08-12 14:00:00".to_string(),
+            "2024-09-13 15:00:00".to_string(),
+            "2025-10-14 16:00:00".to_string(),
+            "2025-13-14 16:00:00".to_string()
+        ],
+        "date" => &[
+            "2023-08-12".to_string(),
+            "2024-09-13".to_string(),
+            "2025-10-14".to_string(),
+            "2025-13-14".to_string()
```
Should you test behavior with bad data? Ill-formated, day-of-month out of range, overflow, underflow.

mccalluc reviewed (commented):
Main concern is how ill-formed / out-of-range dates get handled.

Shoeboxam commented:
I got rid of the fixtures, rebased on top of the refactored ExprDomain, and updated the tests to edge cases per your suggestions @mccalluc. Thanks for the thoughtful review!

mccalluc commented on rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/mod.rs:
```diff
@@ -0,0 +1,115 @@
+    ) {
+        return fallible!(
+            MakeTransformation,
+            "str.strptime output dtype must be Time, micro- or milli-second Datetime or Date, found {}",
```
The comment above also mentions nanoseconds. Add to this message?

mccalluc replied:
Oh, the absence of nano is probably related to https://github.com/pola-rs/polars/issues/19928. Maybe file an issue in this repo to track that?

Shoeboxam replied:
I left an error condition in the code.

mccalluc commented on rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs:
```diff
@@ -0,0 +1,162 @@
+        lit("null"),
+    ));
+
+    println!("{:?}", lf.collect());
```
Stricter test here?

mccalluc commented on rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs:
```diff
@@ -0,0 +1,162 @@
+#[test]
+fn test_strptime_datetime() -> Fallible<()> {
+    let lf = df!("datetime" => &[
+        "+262142-01-01T00:00:00Z",
+        "+262143-01-01T00:00:00Z",
+        "+262144-01-01T00:00:00Z",
```
These are unusual years: Any reason to pick these particular three? I feel like it's trying to exercise something particular, but I'm not sure what.

mccalluc commented on rust/src/transformations/make_stable_expr/namespace_str/expr_strptime/test.rs:
```diff
@@ -0,0 +1,162 @@
+            lit("null"),
+        ));
+
+    let expected = DataFrame::new(vec![Series::new("t", vec![None::<NaiveTime>])])?;
+    assert_eq!(lf.collect()?, expected);
```
If I'm reading this correctly, there's no data in the expected frame. Is this because there's a problem with the decimal seconds? This behavior is really different from the integer seconds that follow. Consider splitting this into two tests, and either in the test name or a comment, explain why this one is different.

mccalluc commented on rust/src/transformations/make_stable_expr/namespace_str/mod.rs:
```diff
@@ -0,0 +1,52 @@
+        #[cfg(feature = "contrib")]
+        StringFunction::Strptime(_, _) => expr_strptime::make_expr_strptime(input_domain, input_metric, expr),
+
+        expr => fallible!(
+            MakeTransformation,
+            "Expr is not recognized at this time: {:?}. {}If you would like to see this supported, please file an issue.",
+            expr,
+            get_disabled_features_message()
+        )
```
This error seems to be conflating two separate things. Maybe they just don't have `contrib` enabled? Or maybe even with `contrib`, it doesn't match?

Maybe have two separate `fallible!`s, one guarded with `#[cfg(feature = "contrib")]`, and one with the negation?

Or if it's better to keep it simple, one error message that makes it clear that there are two different possible root causes?

Shoeboxam replied:
This is what the `get_disabled_features_message()` call does. If there are disabled features it interpolates that explanation into the error message.

mccalluc reviewed (commented):
Thanks for the explanations. Reviewed from the top, and there are still some things that seem odd to me, but I might be confused.

Thinking about scope: I'm not sure if this would be exposed on the python side with this PR, or if there is follow-up work that will do that. If there isn't a follow-up, then it would be nice to have some tests on the Python side, even if we're holding off on docs until the feature set for this release is stabilized.

mccalluc reviewed (approved):
Just looking at the latest changes, and the explanations are very helpful!
